File: android/src/main/java/com/cameravisionpixelcolors/PixelColorsFrameProcessor.kt
package com.cameravisionpixelcolors

import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.VisionCameraProxy

class PixelColorsFrameProcessor(proxy: VisionCameraProxy, options: Map<String, Any>?) : FrameProcessorPlugin() {
  override fun callback(frame: Frame, arguments: Map<String, Any>?): Any {
    val image = frame.image ?: return emptyMap<String, Any>()
    val bitmap = YuvToBitmapConverter.convert(image)

    // Parse options from arguments
    val analysisOptions = parseOptions(arguments)

    PixelAnalyzerEngine.analyzeAsync(bitmap, analysisOptions)
    return PixelAnalyzerEngine.analyzeSync()
  }

  private fun parseOptions(arguments: Map<String, Any>?): AnalysisOptions {
    val optionsMap = arguments?.get("options") as? Map<*, *> ?: return AnalysisOptions()

    val enableMotionDetection = optionsMap["enableMotionDetection"] as? Boolean ?: false
    val motionThreshold = (optionsMap["motionThreshold"] as? Number)?.toFloat() ?: 0.1f

    val roiMap = optionsMap["roi"] as? Map<*, *>
    val roi = if (roiMap != null) {
      val x = (roiMap["x"] as? Number)?.toFloat() ?: 0f
      val y = (roiMap["y"] as? Number)?.toFloat() ?: 0f
      val width = (roiMap["width"] as? Number)?.toFloat() ?: 1f
      val height = (roiMap["height"] as? Number)?.toFloat() ?: 1f
      ROIConfig(x, y, width, height)
    } else null

    val maxTopColors = (optionsMap["maxTopColors"] as? Number)?.toInt() ?: 3
    val maxBrightestColors = (optionsMap["maxBrightestColors"] as? Number)?.toInt() ?: 3

    return AnalysisOptions(
      enableMotionDetection = enableMotionDetection,
      motionThreshold = motionThreshold,
      roi = roi,
      maxTopColors = maxTopColors,
      maxBrightestColors = maxBrightestColors
    )
  }
}

File: ios/PixelColorsFrameProcessor.swift
import VisionCamera
import CoreMedia

@objc(PixelColorsFrameProcessor)
public final class PixelColorsFrameProcessor: FrameProcessorPlugin {
  public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable : Any]! = [:]) {
    super.init(proxy: proxy, options: options)
  }

  public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable : Any]?) -> Any {
    guard let pixelBuffer = CMSampleBufferGetImageBuffer(frame.buffer) else {
      return PixelAnalyzerEngine.shared.analyzeSync()
    }

    // Parse options from arguments
    let options = parseOptions(from: arguments)

    // Fire & forget GPU pipeline
    PixelAnalyzerEngine.shared.analyzeAsync(pixelBuffer: pixelBuffer, options: options)
    // Return latest cached result synchronously
    return PixelAnalyzerEngine.shared.analyzeSync()
  }

  private func parseOptions(from arguments: [AnyHashable: Any]?) -> AnalysisOptionsNative {
    var options = AnalysisOptionsNative()

    guard let args = arguments,
          let optionsDict = args["options"] as? [String: Any] else {
      return options
    }

    if let enableMotionDetection = optionsDict["enableMotionDetection"] as? Bool {
      options.enableMotionDetection = enableMotionDetection
    }

    if let motionThreshold = optionsDict["motionThreshold"] as? Double {
      options.motionThreshold = Float(motionThreshold)
    }

    if let roiDict = optionsDict["roi"] as? [String: Any],
       let x = roiDict["x"] as? Double,
       let y = roiDict["y"] as? Double,
       let width = roiDict["width"] as? Double,
       let height = roiDict["height"] as? Double {
      options.roi = (x: Float(x), y: Float(y), width: Float(width), height: Float(height))
    }

    return options
  }
}
